pyuvdata/uvcal/calh5.py

```diff
@@ -0,0 +1,891 @@
+            spw_inds,
+            freq_inds,
+            jones_inds,
+            history_update_string,
+        ) = self._select_preprocess(
+            antenna_nums=antenna_nums,
+            antenna_names=antenna_names,
+            frequencies=frequencies,
+            freq_chans=freq_chans,
+            spws=spws,
+            times=times,
+            time_range=time_range,
+            lsts=lsts,
+            lst_range=lst_range,
+            jones=jones,
+        )
+        # figure out which axis is the most selective
+        if ant_inds is not None:
+            ant_frac = len(ant_inds) / float(self.Nants_data)
+        else:
+            ant_frac = 1
+
+        if time_inds is not None:
+            time_frac = len(time_inds) / float(self.Ntimes)
+        else:
+            time_frac = 1
+
+        if freq_inds is not None and not self.wide_band:
+            freq_frac = len(freq_inds) / float(self.Nfreqs)
+        else:
+            freq_frac = 1
+
+        if spw_inds is not None and self.wide_band:
+            spw_frac = len(spw_inds) / float(self.Nspws)
+        else:
+            spw_frac = 1
+
+        if jones_inds is not None:
+            jones_frac = len(jones_inds) / float(self.Njones)
+        else:
+            jones_frac = 1
+
+        min_frac = np.min([ant_frac, time_frac, freq_frac, jones_frac, spw_frac])
+
+        quality_present = False
+        if "qualities" in dgrp:
+            quality_present = True
+        total_quality_present = False
+        if "total_qualities" in dgrp:
+            total_quality_present = True
+
+        if min_frac == 1:
+            # no select, read in all the data
+            inds = (np.s_[:], np.s_[:], np.s_[:], np.s_[:])
+            if self.cal_type == "gain":
+                self.gain_array = uvutils._index_dset(dgrp["gains"], inds)
+            else:
+                self.delay_array = uvutils._index_dset(dgrp["delays"], inds)
+            self.flag_array = uvutils._index_dset(dgrp["flags"], inds)
+            if quality_present:
+                self.quality_array = uvutils._index_dset(dgrp["qualities"], inds)
+            if total_quality_present:
+                tq_inds = (np.s_[:], np.s_[:], np.s_[:])
+                self.total_quality_array = uvutils._index_dset(
+                    dgrp["total_qualities"], tq_inds
+                )
+        else:
+            # do select operations on everything except data_array, flag_array
+            # and nsample_array
+            self._select_by_index(
+                ant_inds=ant_inds,
+                time_inds=time_inds,
+                spw_inds=spw_inds,
+                freq_inds=freq_inds,
+                jones_inds=jones_inds,
+                history_update_string=history_update_string,
+            )
+
+            # determine which axes can be sliced, rather than fancy indexed
+            # max_nslice_frac of 0.1 is just copied from uvh5, not validated
+            if ant_inds is not None:
```

Comment:
This may be something to mark as a TODO for later, but there seems to be a lot of shared logic between this and the UVH5 module that might be worth shoving into utils.

Reply:
Yeah, I thought about it briefly but the axis structures were different enough that I gave up. Happy to add a TODO.